### DISCUSSION TITLE
Open-source Token

### DISCUSSION THREAD

**Discussion by r151809** [iainbullock/tesla-http-proxy-docker#55] [Q&A] [answered]
Hi,

Thank you @iainbullock for working for the common good :)

I tried to follow the instructions to my best, but it's quite likely that I have made multiple mistakes. But the first to steps in "Tesla HTTP Proxy setup" -page seem to work, except for the fact that I do not find client_secret from Tesla's page. Client_id is there to copy, but no client_secret. According to Tesla that's the case if you choose [open source token](https://developer.tesla.com/docs/fleet-api#authentication). I'd guess that this has changed quite lately in Tesla's api..

Assuming this is not a SUE (Stupid User Error), what would be your recommended way of getting forward? 
1. I implent [Open source token](https://developer.tesla.com/docs/fleet-api#open-source-token) based authentication. There is a risk that I fail to complete it.. 
2. Change the registration to non open source
3. Wait for someone else to implement it - do you Iain know if someone has started the work?
4. Discard docker based installation

:)

**Comment by iainbullock** [comment, marked as answer]
Did you follow @tux43 instructions here https://www.smartmotion.life/2024/04/23/tesla-custom-integration-with-home-assistant-on-docker/?v=6cc98ba2045f

In particular section 4.9 shows where you get Client ID and Client Secret 

(Unless something has changed as I haven't had to run this process since my first set up)

Looking at the Tesla documentation, I believe you should be generating a Partner Authentication Token.

Another place to look would be the parent project which has a more active user base (and therefore the same question may have already been asked) compared to my project https://github.com/llamafilm/tesla-http-proxy-addon

Let me know how you get on

**Comment by Olen** [comment]
I noticed the same thing as the OP.  There are now two options you can choose from when creating an App in the Tesla Developer Portal.  If you choose "Open Source" there is no Client Secret generated, so you must make sure you say "No" on this question (even if your app is going to be open source, because you will probably not publish your client-id and secret in the source of the app):

![image](https://github.com/user-attachments/assets/87e5cdff-3b42-4dd7-9cd9-bbc33df2d8b6)

I don't know anything about how auth works on the Open Source version, but since you do not have a client secret, the Oauth-dance is probably quite different from the Non Open Source version.

**Comment by r151809** [comment]
Thank you.
Managed to get further by "requesting" second instance without open source selection.

I still have something with my local dns. At some point tesla_http_proxy did not like my domain (.home) so I tried with .local. Now HA does not resolve .local and .home goes to SSLCertVerificationError. But this is something I will be able to solve somehow - it just takes time.

Luckily it's summer and I do not really need load shredding yet :)

**Reply by iainbullock** [reply]
OK let me know how you progress.

If you are planning to use this so HA can control charging current for solar excess, you may find you get rate limited by Tesla's API. This was the driver behind another project that I'm involved with:

https://github.com/tesla-local-control/tesla_ble_mqtt_docker

It is not a replacement for tesla-http-proxy, but it does let you send a large number of commands via bluetooth without the rate limiting that can occur when using the API.

**Reply by r151809** [reply]
That's awesome - I had totally missed the option of using BLE. It will not work for our 2014 S (excluded from the fleet API anyways), but 2019 Model 3 should be OK.

BLE should depend less on cloud: more reliable and faster :)

Our main fuses are 3x25 A and the house is electrically heated (in Finland). In winter time heating alone might go to the fuse limits, and the cars should take only the reminders, preferably optimizing the electricity price also..

**Reply by r151809** [reply]
Hi,

I'm strugling to get BLE working in my HA server. Host BT does not support BLE, and the dongle I found is some Chinese crap. I managed once to have presence "true", but key provisioning failed always.. I do not want to move the server, and running tesla_ble_mqtt on a RPi is only the last choice. I'll try to find a better BT-dongle first, and reconsider RPi if that fails.

Ideally speaking :)
I have good experience with ESPHome and BT Proxy - you can cover big areas with those quite easily. So someone should port the go-code to ESPHome or change the BLE code to use https://esphome.io/components/ble_client.html .

I'll keep you updated - currently I'm planning not to use fleet API.

**Reply by tux43** [reply]
> Ideally speaking :)
> I have good experience with ESPHome and BT Proxy - you can cover big areas with those quite easily. So someone should port the go-code to ESPHome or change the BLE code to use https://esphome.io/components/ble_client.html .

Does this mean that ESP Bluetooth Proxy will not work? https://esphome.io/projects/?type=bluetooth

**Reply by iainbullock** [reply]
I don't believe ESP Bluetooth Proxy will work as Tesla's Go code wants to talk to a BT device directly. Maybe you could write your own Go code, but I haven't looked at it in any detail.

**Reply by r151809** [reply]
I'm now waiting for my kid in the car, so I had spare time. I tried directly https://github.com/teslamotors/vehicle-command in Mac. It works nicely. So I'm now confident that I will get BLE up and running somehow under HA :) In my case no need for tesla_http_proxy, as tesla_ble_mqtt serves me better :)

Thanks @iainbullock !

**Reply by iainbullock** [reply]
Good to hear. You might still want   tesla_http_proxy if you want more sensors as there's not much with BLE

**Reply by r151809** [reply]
You are probably right - there is no SOC in the BLE data sets..

**Reply by r151809** [reply]
Got tesla_http_proxy running finally.

The issue wat that my HA was not able to find the proxy, and after fixing that it was accessing the internet facing https-port of nginx.

Failed ping between docker containers was due to invalid network_mode configs. The default, bridge, should work in most cases - I had a mess of host and bridge so some things did work, but not all..

Now I have separate network called http_proxy in docker-compose. This way I do not have to worry about conflicting ports. In the setup phase external setting must be true, of course.

```
  tesla_http_proxy:
    networks:
      - http_proxy

  ha:
    networks:
      - http_proxy
      - default

networks:
  http_proxy:
    external: true
```
With this config HA connects to proxy like this: httos://tesla_http_proxy/

And one note for other who make mistakes in the configs and want to start from the beginning: r
* Removing access_token enables Flask
* Removing key.pem forces selfsigned certificate renewal

Thank you @iainbullock and @Olen :)
